packages/server/modules/workspaces/repositories/workspaces.ts

```diff
@@ -35,26 +38,92 @@ export const upsertWorkspaceFactory =
       .merge(['description', 'logoUrl', 'name', 'updatedAt'])
   }
 
-export const getWorkspaceRoleFactory =
-  ({ db }: { db: Knex }): GetWorkspaceRole =>
+export const getWorkspaceRolesFactory =
+  ({ db }: { db: Knex }): GetWorkspaceRoles =>
+  async ({ workspaceId }) => {
+    return await tables.workspacesAcl(db).select('*').where({ workspaceId })
+  }
+
+export const getWorkspaceRoleForUserFactory =
+  ({ db }: { db: Knex }): GetWorkspaceRoleForUser =>
   async ({ userId, workspaceId }) => {
-    const acl = await tables
-      .workspacesAcl(db)
-      .select('*')
-      .where({ userId, workspaceId })
-      .first()
+    return (
+      (await tables
+        .workspacesAcl(db)
+        .select('*')
+        .where({ userId, workspaceId })
+        .first()) ?? null
+    )
+  }
 
-    return acl || null
+export const getWorkspaceRolesForUserFactory =
+  ({ db }: { db: Knex }): GetWorkspaceRolesForUser =>
+  async ({ userId }, options) => {
+    const workspaceIdFilter = options?.workspaceIdFilter ?? []
+
+    const query = tables.workspacesAcl(db).select('*').where({ userId })
+
+    if (workspaceIdFilter.length > 0) {
+      query.whereIn('workspaceId', workspaceIdFilter)
+    }
+
+    return await query
+  }
+
+const isUserLastWorkspaceAdmin = (
+  workspaceRoles: WorkspaceAcl[],
+  userId: string
+): boolean => {
+  const workspaceAdmins = workspaceRoles.filter(
+    ({ role }) => role === 'workspace:admin'
+  )
+  const isUserAdmin = workspaceAdmins.some((role) => role.userId === userId)
+
+  return isUserAdmin && workspaceAdmins.length === 1
+}
+
+export const deleteWorkspaceRoleFactory =
+  ({ db }: { db: Knex }): DeleteWorkspaceRole =>
+  async ({ userId, workspaceId }) => {
+    // Protect against removing last admin in workspace
+    const workspaceRoles = await getWorkspaceRolesFactory({ db })({ workspaceId })
+
+    if (isUserLastWorkspaceAdmin(workspaceRoles, userId)) {
```

Comment:
There could be cases, where we want to delete the role, even if the user is the last admin ie workspace deletion. ( If its not done via cascading deletes for some reason ). That way, this function becomes even more complex, cause based on an external logic, we might want to allow deletion after all. Its a lot better to have this check in the service implementation.